### PR TITLE
Simplify VpiArrayObjHdl::initialise

### PR DESCRIPTION
**Pull request by AndrewNolte**
I had issues with range_idx being calculated incorrectly, so I wanted to see what simulators actually require this, if at all. It gets calculated incorrectly when using get_handle_by_name() with a fully qualified name (not from handle.py) with brackets in other parts of the path. I didn't see range_idx being actually used in testing on Xcelium.  It doesn't look like there were any tests when it was put in https://github.com/cocotb/cocotb/commit/54d2965dc29f226f3971ca6e5b6b3cc67149fbc1

### CODE REVIEW

File: src/cocotb/share/lib/vpi/VpiCbHdl.cpp
Comment (ktbarrett) [inline src/cocotb/share/lib/vpi/VpiCbHdl.cpp:132]: Not sure why these were moved up, this is less readable. Variable declarations should be as close to their use as possible.

**Review by ktbarrett — COMMENTED** [review]
Does this fix your issue simply because only Questa/ModelSim, which you aren't using, need pseudo-regions, which you've segregated into ifdef's? So the problem of supporting `get_handle_by_name` from the root on Questa, if there are multiple indexes, is still present, correct? But this commit removes this logic, which isn't necessary for the other simulators?

What tests exist in the regression which test this?

@marlonjames Do you have any concerns here?

**Comment by AndrewNolte** [comment]
> s_vpi_value

Yeah we have use cases where we query objects from the root handle, but the logic for searching through the brackets causes issues. Are we okay with adding tests specifically for the simulator module? I think a good test case would be iterating through handles in a large design and asserting that subhdl == roothdl.get_handle_by_name(subhdl._path), and can be skipped on questa for now.

**Comment by ktbarrett** [comment]
> Are we okay with adding tests specifically for the simulator module?

Oh absolutely! I encourage it.

> can be skipped on questa for now.

That's fine. We can maybe open an upstream ticket or at least track it in an issue if you could open that with the particulars that I'm not yet groking.

**Comment by AndrewNolte** [comment]
I added it to the iteration test case, but ideally it would test on all of the designs, with multiple arrays in the path. Also, while looking at the iteration tests, I think I found that these test tags were removed on accident? https://github.com/cocotb/cocotb/commit/db1cce76e63a52bb7b4c06db39a767199c54092c#diff-1082c9445dd85ec000286580d8e757077f81782e15cd253a1279b956a3cad2cb

**Comment by marlonjames** [comment]
I will try to take a look at this soon.

**Comment by ktbarrett** [comment]
Unfortunately I don't think there's anything we can do until the private CI job hang is resolved. So sorry if we don't look at this sooner.

**Comment by imphil** [comment]
> Unfortunately I don't think there's anything we can do until the private CI job hang is resolved. So sorry if we don't look at this sooner.

That's resolved now, sorry for the disruption.

File: tests/test_cases/test_iteration_verilog/test_iteration_es.py
Comment (ktbarrett) [inline tests/test_cases/test_iteration_verilog/test_iteration_es.py:80]: So just to get this straight, this is testing that the object retrieved by name from the root object is the same object as the one found during iteration-based discovery? And that is not the case for Icarus and Riviera for some reasons... A comment might be worthwhile.

The above C code changes look good.

Comment (AndrewNolte) [reply tests/test_cases/test_iteration_verilog/test_iteration_es.py:80]: Yeah, but I might've mistaked the CI being out of commission as a test failure?

Comment (ktbarrett) [reply tests/test_cases/test_iteration_verilog/test_iteration_es.py:80]: Well it was just re-run and seems to be working now.

Comment (AndrewNolte) [reply tests/test_cases/test_iteration_verilog/test_iteration_es.py:80]: ok I'll revert and push to see if it works on those sims

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
@marlonjames Could you have a look?

File: src/cocotb/share/lib/vpi/VpiCbHdl.cpp
Comment (marlonjames) [inline src/cocotb/share/lib/vpi/VpiCbHdl.cpp:171]: ?
```suggestion
        LOG_ERROR("Unable to get range for indexable array or memory");
```

File: src/cocotb/share/lib/vpi/VpiCbHdl.cpp
Comment (marlonjames) [inline src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: `std::string hdl_name = vpi_get_str(vpiName, hdl);` should be pulled out of the `#ifdef`, and the non-Modelsim case should enforce that `hdl_name == name` so that we get an error if there is another VPI simulator that needs pseudo-regions for multi-dimensional arrays.

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: I'm don't think name != hdl_name always implies this? In Verilator when there's a generate scope, hdl_name = 'my_mod.scope' and name = 'scope', but maybe that's another issue for fix on the Verilator side...

Comment (marlonjames) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: `vpiName` should be the local name of the object, `vpiFullName` should be the full hierarchical name.

Comment (marlonjames) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Because pseudo-regions use the handle of the parent (there is no proper VPI object with a handle), the names should mismatch.
That is how it works now. Pseudo-regions should be re-worked to preserve more context, right now the checking for and handling of pseudo-regions is scattered across the GPI codebase.

Comment (ktbarrett) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Maybe this should be a follow on then? One thing at a time with a PR. Maybe open an issue for future investigation.

Comment (marlonjames) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Then I'd rather special-case the simulator(s) that have a problem with the pseudo-region check, instead of the other way around.
I see in Verilator's code that for scopes, there isn't a difference between `vpiName` and `vpiFullName`, so that is likely the cause of the problem here. I might be okay with special-casing for Verilator until that is fixed, but like #3664 I think the change belongs in Verilator.

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Yeah I think verilator needs a test case for just iterating through the tree and printing the handle info. Maybe we can add that to cocotb to measure simulator vpi compatibility, since the infrastructure for running different sims is already there? Or maybe that would belong on https://chipsalliance.github.io/sv-tests-results/

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: The original issue for me was that in create_gpi_obj_from_handle, name is assumed to be the vpiName of the object, but it's not the case if calling the get_handle_by_name from a handle that isn't the direct parent handle. Would it be possible to add some flag there to not create pseudo handles, or not create them if we're querying down further than just the children?

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Hey following up on this, if the pseudo handle check here is only needed for modelsim, wouldn't we rather have it only create handles for that sim? 

Also, I made the vpi dump verilator test case, and it looks like the scopes are generated, so we may just need to add pseudo-handle code specific to verilator https://github.com/verilator/verilator/pull/4838/files

Comment (marlonjames) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Ok, I think I have a better understanding now.
So your problem came from passing a hierarchy to `SimHandleBase._handle.get_handle_by_name()` rather than just the name of a direct child object.

There is no analysis of the `name` passed into `native_check_create()`, it is just concatenated with the name of the `parent` handle and looked up via `vpi_handle_by_name()`. The possibly-a-hierarchy name is then passed through to `create_gpi_obj_from_handle()`, which also has the assumption that it is the name of the object. Finally, that is passed to `initialise()` as if it is the name.

I think we need to either fix the code making the assumptions, or provide a separate function to call when looking up a handle that is not a direct child (e.g. we recommend using `_id()` with a hierarchical path for some cases where simulators have problems with generates etc).

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiCbHdl.cpp:129]: Yeah I did some more testing and this ifdef should actually be included. The one below does seem to be verify specific to modelsim though

**Comment by marlonjames** [comment]
I'll take another look at this soon. Please don't merge until then.

**Comment by ktbarrett** [comment]
Re-running CI.

File: tests/test_cases/test_iteration_verilog/test_iteration_es.py
Comment (ktbarrett) [inline tests/test_cases/test_iteration_verilog/test_iteration_es.py:58]: `cocotb.LANGUAGE` was removed.

**Review by ktbarrett — APPROVED** [review]
LGTM. @marlonjames?

File: src/cocotb/share/lib/vpi/VpiObj.cpp
Comment (marlonjames) [inline src/cocotb/share/lib/vpi/VpiObj.cpp:79]: If `rangeHdl` is null, the iterator handle is no longer valid so must not be freed here.

File: src/cocotb/share/lib/vpi/VpiObj.cpp
Comment (marlonjames) [inline src/cocotb/share/lib/vpi/VpiObj.cpp:71]: This code isn't Modelsim specific.
The problem in `recursive_discovery` test has to do with `dut._handle.get_handle_by_name(subpath)` as detailed in #3710.

Comment (AndrewNolte) [reply src/cocotb/share/lib/vpi/VpiObj.cpp:71]: If it's not modelsim specific, why isn't there a testcase in the other sims that require this codepath? Modelsim was the only one that failed when I took this out, and taking this logic out sovled #3710 for me.

File: src/cocotb/share/lib/vpi/VpiObj.cpp
Comment (ktbarrett) [inline src/cocotb/share/lib/vpi/VpiObj.cpp:75]: I think this line is a different error. We *can* scan, but it's yielding less than the identified number of dimensions. So instead of "Unable to get range for indexable array", maybe it's "Unable to get range for all dimensions of indexable array"?

File: src/cocotb/share/lib/vpi/VpiObj.cpp
Comment (ktbarrett) [reply src/cocotb/share/lib/vpi/VpiObj.cpp:71]: I think the issue is that conceptually this code should *always* run. If we get a multidimensional array, such that `range_idx` is non-0, this code should always be run to get the correct range spec.

Perhaps the only reason that Questa hits this line is because Questa is the only sim that *tests* this line. Might this have to do with how pseudo-regions are handled?

**Review by ktbarrett — CHANGES_REQUESTED** [review]
Now that I better understand the context, I see a better solution. 

We need to get the number of dimensions if the object is a pseudo-region. What we need to do is find only the trailing indexes *that aren't a part of the handle*. The problem is when objects are attempts to be created not from parents, but from higher in the hierarchy. So what needs to be done is:

```
std::string hdl_name = vpi_get(vpiName, hdl);
std::string idx_str = name.substr(name.find_last_of(hdl_name)+hdl_name.length());
std::count(idx_str.begin(), idx_str.end(), '[');
```

Then *unconditionally* do the part in the MODELSIM ifdef.

File: src/cocotb/share/lib/vpi/VpiObj.cpp
Comment (ktbarrett) [inline src/cocotb/share/lib/vpi/VpiObj.cpp:59]:
```suggestion
        range_idx = (int)std::count(&name.begin()[idx_str], name.end(), '[');
```

File: tests/test_cases/test_iteration_verilog/test_iteration_es.py
Comment (ktbarrett) [inline tests/test_cases/test_iteration_verilog/test_iteration_es.py:50]: Commented out code.

**Review by ktbarrett — APPROVED** [review]
Looks good now.

**Comment by ktbarrett** [comment]
The failing test is unrelated. A combination of PRs that passed their tests separately don't pass together after merging. Whoops...

**Pull request merged**
